test/e2e/network/topology_hints.go

```diff
@@ -0,0 +1,201 @@
+
+		var finalSlices []discoveryv1.EndpointSlice
+		err = wait.Poll(5*time.Second, 3*time.Minute, func() (bool, error) {
+			slices, listErr := c.DiscoveryV1().EndpointSlices(f.Namespace.Name).List(context.TODO(), metav1.ListOptions{LabelSelector: fmt.Sprintf("%s=%s", discoveryv1.LabelServiceName, svc.Name)})
+			if listErr != nil {
+				return false, listErr
+			}
+
+			numEndpoints := 0
+			for _, slice := range slices.Items {
+				numEndpoints += len(slice.Endpoints)
+			}
+			if len(nodeNames) > numEndpoints {
+				framework.Logf("Expected %d endpoints, got %d", len(nodeNames), numEndpoints)
+				return false, nil
+			}
+
+			finalSlices = slices.Items
+			return true, nil
+		})
+		framework.ExpectNoError(err, "timed out waiting for EndpointSlices to be ready")
+
+		ginkgo.By("having hints set for each endpoint")
+		for _, slice := range finalSlices {
+			for _, ep := range slice.Endpoints {
+				if ep.Zone == nil {
+					framework.Failf("Expected endpoint in %s to have zone: %v", slice.Name, ep)
+				}
+				if ep.Hints == nil || len(ep.Hints.ForZones) == 0 {
+					framework.Failf("Expected endpoint in %s to have hints: %v", slice.Name, ep)
+				}
+				if len(ep.Hints.ForZones) > 1 {
+					framework.Failf("Expected endpoint in %s to have exactly 1 zone hint, got %d: %v", slice.Name, len(ep.Hints.ForZones), ep)
+				}
+				if *ep.Zone != ep.Hints.ForZones[0].Name {
+					framework.Failf("Expected endpoint in %s to have same zone hint, got %s: %v", slice.Name, *ep.Zone, ep)
+				}
+			}
+		}
+
+		nodeList, err := c.CoreV1().Nodes().List(context.TODO(), metav1.ListOptions{})
+		framework.ExpectNoError(err)
+		nodesByZone := map[string]string{}
+		for _, node := range nodeList.Items {
+			if zone, ok := node.Labels[v1.LabelTopologyZone]; ok {
+				nodesByZone[node.Name] = zone
+			}
+		}
+
+		podList, err := c.CoreV1().Pods(f.Namespace.Name).List(context.TODO(), metav1.ListOptions{})
+		framework.ExpectNoError(err)
+		podsByZone := map[string]string{}
+		for _, pod := range podList.Items {
+			if zone, ok := nodesByZone[pod.Spec.NodeName]; ok {
+				podsByZone[pod.Name] = zone
+			}
+		}
+
+		ginkgo.By("keeping requests in the same zone")
+		for i, nodeName := range nodeNames {
+			// Iterate through max of 3 nodes
+			if i > 2 {
+				break
+			}
+			fromZone, ok := nodesByZone[nodeName]
+			if !ok {
+				framework.Failf("Expected zone to be specified for %s node", nodeName)
+			}
+			ginkgo.By("creating a client pod for probing the service from " + fromZone)
+			podName := "curl-from-" + fromZone
+			clientPod := e2epod.NewAgnhostPod(f.Namespace.Name, podName, nil, nil, nil, "serve-hostname")
+			nodeSelection := e2epod.NodeSelection{Name: nodeName}
+			e2epod.SetNodeSelection(&clientPod.Spec, nodeSelection)
+			cmd := fmt.Sprintf(`date; for i in $(seq 1 3000); do sleep 1; echo "Date: $(date) Try: ${i}"; curl -q -s --connect-timeout 2 http://%s:80/ ; echo; done`, svc.Name)
```

Comment:
does it server the hostname in the root url or in the `/hostname` ?

Reply:
Serves pod name with current request.